### PR TITLE
Enable Exporting Leases to JSON

### PR DESCRIPTION
**Pull request by psychomario**
Handles `SIG{INT,ALRM,TERM,HUP}` to dump out the leases object to `--saveleases` file as json. On start with `--saveleases` the `self.leases` object is updated with the contents, so non-expired ~~(__BUG__ we don't check, will fix)~~ (fixed in 2936162) leases are used

The keys are translated to human readable (and json compatible) mac addresses on export, and back on import.

I've also moved `self.leases[client_mac]['options']` to a separate object. It means we don't have to clean up the `self.leases` dict every time we save it.

### CODE REVIEW

File: pypxe/dhcp.py [hunk 1]
```diff
@@ -47,6 +48,7 @@ def __init__(self, **server_settings):
         self.mode_verbose = server_settings.get('mode_verbose', False) # debug mode
         self.mode_debug = server_settings.get('mode_debug', False) # debug mode
         self.logger = server_settings.get('logger', None)
+        self.saveleasesfile = server_settings.get('saveleases', '')
```
Comment (mmattioli) [inline pypxe/dhcp.py]: perhaps use `self.save_leases_file`?

File: README.md [hunk 2]
```diff
@@ -55,6 +55,7 @@ The following are arguments that can be passed to `pypxe.server` when running fr
 |__`--verbose`__|Enable selected services in verbose mode; services are selected by passing the name in a comma separated list. Follows the same syntax for selecting service as `--debug`._|`''`|
 |__`--config`__|Load configuration from JSON file. (see [`example_cfg.json`](example_cfg.json))|`None`|
 |__`--static-config`__|Load DHCP lease configuration from JSON file. (see [`example-leases.json`](example-leases.json))|`None`|
+|__`--saveleases`|Saves allocated leases on `HUP`, `INT`(`^C`), `ALRM` and `TERM`. Loads on start if the file exists already|`None`|
```
Comment (mmattioli) [inline README.md]: perhaps use `--save-leases`?

**Comment by mmattioli** [comment]
@psychomario looks like you didn't use `pickle`? Also, this looks good but I just wanted to point out the variable and function naming should maybe be more inline with PEP8 (`def export_leases` rather than `def exportLeases`).

**Comment by psychomario** [comment]
I don't really like `pickle`, it leaves a completely opaque binary file on disk, it would have worked significantly better for storing the binary data from `options`, but given that I've moved that elsewhere it isn't an issue anymore.. This method leaves a nice json file which, for whatever reason, is editable and viewable.

**Comment by psychomario** [comment]
I've pushed a new commit with the style improvements you suggested.

File: pypxe/dhcp.py [hunk 3]
```diff
@@ -98,7 +100,39 @@ def __init__(self, **server_settings):
         self.sock.bind(('', self.port ))
 
         # key is MAC
+        # separate options dict so we don't have to clean up on export
+        self.options = dict()
         self.leases = defaultdict(lambda: {'ip': '', 'expire': 0, 'ipxe': self.ipxe})
+        if self.save_leases_file:
+            try:
+                leasesfile = open(self.save_leases_file, 'rb')
```
Comment (mmattioli) [inline pypxe/dhcp.py]: `leases_file`

File: pypxe/dhcp.py [hunk 4]
```diff
@@ -98,7 +100,39 @@ def __init__(self, **server_settings):
...
+                imported = json.load(leasesfile)
+                importsafe = dict()
```
Comment (mmattioli) [inline pypxe/dhcp.py]: `import_safe`

File: pypxe/dhcp.py [hunk 5]
```diff
@@ -98,7 +100,39 @@ def __init__(self, **server_settings):
...
+                for lease in imported:
+                    packedmac = struct.pack('BBBBBB', *map(lambda x:int(x, 16), lease.split(':')))
```
Comment (mmattioli) [inline pypxe/dhcp.py]: `packed_mac`

**Comment by mmattioli** [comment]
@psychomario sorry to be a pest but I made a few line comment where variable names should still be modified to conform with PEP8. Give it a look once more? Thanks.

**Comment by psychomario** [comment]
As ever, you're correct. I'm not able to make these changes for the next few days but I will do so when I can.

**Comment by mmattioli** [comment]
@psychomario since #111 was just merged you're going to need to do a rebase anyway lol

**Comment by psychomario** [comment]
@mmattioli I think I've got all your changes, I almost made a mess of the branch in the process.

**Comment by mmattioli** [comment]
:+1: LGTM

**Pull request merged**